api/app/weather_models/rdps_filename_marshaller.py

```diff
@@ -77,28 +98,49 @@ def parse_rdps_filename(url: str):
     return (forecast_start_date, run_hour, forecast_hour)
 
 
-def check_compose_invariants(forecast_start_date: datetime, run_hour: int, forecast_hour: int):
+def check_compose_invariants(forecast_start_date: datetime, run_hour: int, forecast_hour: int, weather_parameter: str):
     """Explode if any of these assertions fail"""
     assert forecast_start_date.tzinfo is not None
     assert int(forecast_start_date.utcoffset().total_seconds()) == 0
     assert f"{forecast_hour:03d}" in FORECAST_HOURS
     assert run_hour in list(range(0, 36))
+    assert weather_parameter in weather_key_parameters
 
 
-def compose_computed_rdps_filename(forecast_start_date: datetime, run_hour: int, forecast_hour: int, source_prefix: Literal[SourcePrefix.CMC, SourcePrefix.COMPUTED]):
+def compose_rdps_filename(forecast_start_date: datetime, run_hour: int, forecast_hour: int, weather_parameter: str):
     """Compose and return a computed RDPS url given a forecast start date, run hour and forecast hour."""
-    check_compose_invariants(forecast_start_date, run_hour, forecast_hour)
+    check_compose_invariants(forecast_start_date, run_hour, forecast_hour, weather_parameter)
+    key_params = get_weather_key_params(weather_parameter)
     model_hour = model_run_for_hour(run_hour)
     adjusted_forecast_hour = adjust_forecast_hour(run_hour, forecast_hour)
-    file_ext = ".grib2" if source_prefix == SourcePrefix.CMC else ".tif"
+    file_ext = ".grib2"
 
     return (
-        f"{source_prefix.value}{DELIMITER}{REG}{DELIMITER}{APCP}{DELIMITER}{SFC}{DELIMITER}{LEVEL}{DELIMITER}{PS10KM}{DELIMITER}"
+        f"{SourcePrefix.CMC.value}{DELIMITER}{REG}{DELIMITER}{key_params.variable}{DELIMITER}{key_params.level_type}{DELIMITER}{key_params.level}{DELIMITER}{PS10KM}{DELIMITER}"
         f"{forecast_start_date.date().isoformat().replace('-','')}{model_hour:02d}{DELIMITER}P{adjusted_forecast_hour:03d}{file_ext}"
     )
 
 
-def compose_computed_precip_rdps_key(forecast_start_date: datetime, run_hour: int, forecast_hour: int, source_prefix: Literal[SourcePrefix.CMC, SourcePrefix.COMPUTED]):
+def compose_precip_rdps_key(forecast_start_date: datetime, run_hour: int, forecast_hour: int):
     """Compose and return a computed RDPS url given a forecast start date, run hour and forecast hour."""
     model_hour = model_run_for_hour(run_hour)
-    return f"{model_hour:02d}/precip/{compose_computed_rdps_filename(forecast_start_date, run_hour, forecast_hour, source_prefix)}"
+    return f"{model_hour:02d}/precip/{compose_rdps_filename(forecast_start_date, run_hour, forecast_hour, 'precip')}"
+
+
+def compose_computed_rdps_filename(accumulation_end_datetime: datetime):
+    """Compose and return a computed RDPS url given the datetime that precip is being accumulated to."""
+    key_params = get_weather_key_params("precip")
+    model_hour = model_run_for_hour(accumulation_end_datetime.hour)
+    adjusted_forecast_hour = accumulation_end_datetime.hour - model_hour
```

Comment:
I think we may still have the negative hour bug here since `accumulation_end_datetime` is based off the current time the job runs + an offset hour. RDPS jobs run every hour I believe to make up for failed jobs so we could have a current_time + hour offset combo that results in a negative value. I think what we want to do is simplify the `adjusted_forecast_hour` to just be the offset hour from the model hour.

@dgboss thoughts?

Reply:
I'll step through the code after lunch, but I think you are correct.

Reply:
I did step through and test that, it won't create those negative hours anymore. Just did it again for a sanity check.

`main`
CMC_reg_APCP_SFC_0_ps10km_2024072912_P009.grib2
CMC_reg_APCP_SFC_0_ps10km_2024072912_P033.grib2
COMPUTED_reg_APCP_SFC_0_ps10km_2024073012_P-11.tif

`task/rdps-24hr-precip-output`
CMC_reg_APCP_SFC_0_ps10km_2024072912_P009.grib2
CMC_reg_APCP_SFC_0_ps10km_2024072912_P033.grib2
COMPUTED_reg_APCP_SFC_0_ps10km_2024073012_P009.tif

Doesn't necessarily mean that's the naming convention we want though I suppose